shell/gpu/gpu_surface_metal.mm

```diff
@@ -93,65 +95,52 @@
-                                            1,                      // sample count
-                                            metal_texture_info      // metal texture info
+  auto surface = SkSurface::MakeFromCAMetalLayer(context_.get(),            // context
+                                                 layer_.get(),              // layer
+                                                 kTopLeft_GrSurfaceOrigin,  // origin
+                                                 1,                         // sample count
+                                                 kBGRA_8888_SkColorType,    // color type
+                                                 nullptr,                   // colorspace
+                                                 nullptr,                   // surface properties
+                                                 &next_drawable_  // drawable (transfer out)
   );
 
-  auto command_buffer =
-      fml::scoped_nsprotocol<id<MTLCommandBuffer>>([[command_queue_.get() commandBuffer] retain]);
-
-  SkSurface::RenderTargetReleaseProc release_proc = [](SkSurface::ReleaseContext context) {
-    [reinterpret_cast<id>(context) release];
-  };
-
-  auto surface =
-      SkSurface::MakeFromBackendRenderTarget(context_.get(),            // context
-                                             metal_render_target,       // backend render target
-                                             kTopLeft_GrSurfaceOrigin,  // origin
-                                             kBGRA_8888_SkColorType,    // color type
-                                             nullptr,                   // colorspace
-                                             nullptr,                   // surface properties
-                                             release_proc,              // release proc
-                                             metal_texture.release()    // release context (texture)
-      );
-
   if (!surface) {
     FML_LOG(ERROR) << "Could not create the SkSurface from the metal texture.";
     return nullptr;
   }
 
-  bool hasExternalViewEmbedder = delegate_->GetExternalViewEmbedder() != nullptr;
-
   // External views need to present with transaction. When presenting with
   // transaction, we have to block, otherwise we risk presenting the drawable
   // after the CATransaction has completed.
   // See:
   // https://developer.apple.com/documentation/quartzcore/cametallayer/1478157-presentswithtransaction
   // TODO(dnfield): only do this if transactions are actually being used.
   // https://github.com/flutter/flutter/issues/24133
-  auto submit_callback = [drawable = next_drawable, command_buffer, hasExternalViewEmbedder](
-                             const SurfaceFrame& surface_frame, SkCanvas* canvas) -> bool {
+  auto submit_callback = [this](const SurfaceFrame& surface_frame, SkCanvas* canvas) -> bool {
     canvas->flush();
-    if (!hasExternalViewEmbedder) {
+
+    if (next_drawable_ == nullptr) {
```

Comment:
Is it possible that this callback gets invoked between lines 101 and 104 above?

Reply:
No. These are on the same thread and tasks are posted serially.